src/scripts/arabic.rs

```diff
@@ -0,0 +1,140 @@
+    [ 0xFE8D, 0xFE8E, NONE, NONE, 0xFEFB, 0xFEFC],
+    [ 0xFE8F, 0xFE90, 0xFE91, 0xFE92, NONE, NONE],
+    [ 0xFE93, 0xFE94, NONE, NONE, NONE, NONE],
+    [ 0xFE95, 0xFE96, 0xFE97, 0xFE98, NONE, NONE],
+    [ 0xFE99, 0xFE9A, 0xFE9B, 0xFE9C, NONE, NONE],
+    [ 0xFE9D, 0xFE9E, 0xFE9F, 0xFEA0, NONE, NONE],
+    [ 0xFEA1, 0xFEA2, 0xFEA3, 0xFEA4, NONE, NONE],
+    [ 0xFEA5, 0xFEA6, 0xFEA7, 0xFEA8, NONE, NONE],
+    [ 0xFEA9, 0xFEAA, NONE, NONE, NONE, NONE],
+    [ 0xFEAB, 0xFEAC, NONE, NONE, NONE, NONE],
+    [ 0xFEAD, 0xFEAE, NONE, NONE, NONE, NONE],
+    [ 0xFEAF, 0xFEB0, NONE, NONE, NONE, NONE],
+    [ 0xFEB1, 0xFEB2, 0xFEB3, 0xFEB4, NONE, NONE],
+    [ 0xFEB5, 0xFEB6, 0xFEB7, 0xFEB8, NONE, NONE],
+    [ 0xFEB9, 0xFEBA, 0xFEBB, 0xFEBC, NONE, NONE],
+    [ 0xFEBD, 0xFEBE, 0xFEBF, 0xFEC0, NONE, NONE],
+    [ 0xFEC1, 0xFEC2, 0xFEC3, 0xFEC4, NONE, NONE],
+    [ 0xFEC5, 0xFEC6, 0xFEC7, 0xFEC8, NONE, NONE],
+    [ 0xFEC9, 0xFECA, 0xFECB, 0xFECC, NONE, NONE],
+    [ 0xFECD, 0xFECE, 0xFECF, 0xFED0, NONE, NONE],
+    [ 0, 0, NONE, NONE, NONE, NONE],
+    [ 0, 0, NONE, NONE, NONE, NONE],
+    [ 0, 0, NONE, NONE, NONE, NONE],
+    [ 0, 0, NONE, NONE, NONE, NONE],
+    [ 0, 0, NONE, NONE, NONE, NONE],
+    [ 0x640, 0x640, 0x640, 0x640, NONE, NONE],
+    [ 0xFED1, 0xFED2, 0xFED3, 0xFED4, NONE, NONE],
+    [ 0xFED5, 0xFED6, 0xFED7, 0xFED8, NONE, NONE],
+    [ 0xFED9, 0xFEDA, 0xFEDB, 0xFEDC, NONE, NONE],
+    [ 0xFEDD, 0xFEDE, 0xFEDF, 0xFEE0, NONE, NONE],
+    [ 0xFEE1, 0xFEE2, 0xFEE3, 0xFEE4, NONE, NONE],
+    [ 0xFEE5, 0xFEE6, 0xFEE7, 0xFEE8, NONE, NONE],
+    [ 0xFEE9, 0xFEEA, 0xFEEB, 0xFEEC, NONE, NONE],
+    [ 0xFEED, 0xFEEE, NONE, NONE, NONE, NONE],
+    [ 0xFEEF, 0xFEF0, NONE, NONE, NONE, NONE],
+    [ 0xFEF1, 0xFEF2, 0xFEF3, 0xFEF4, NONE, NONE],
+];
+
+fn get_contextual_form_of_char(prev: char, next: char, c: char) -> Option<u32> {
+    let cp = c as u32;
+    let next_cp = next as u32;
+    let prev_cp = prev as u32;
+
+    if !is_arabic_letter(cp) {
+        return Some(cp); // Not an Arabic letter, so return it as is.
+    }
+
+    let is_la = is_lam_alif(cp, next_cp);
+    let is_apl = is_alif_after_lam(prev_cp, cp);
+    let is_lapl = is_la || is_apl;
+
+    // The reference index now takes into account the special Lam-Alif combination.
+    let ref_index = if is_la { next_cp } else { cp } - AR_START;
+
+    let target = ((((is_lapl || is_arabic_letter(next_cp)) && is_linking_type(cp)) as u32) << 1)
+        | is_linking_type(prev_cp) as u32;
+
+    if let Some(form) = ARABIC_FORMS.get(ref_index as usize) {
+        if is_lapl {
+            if is_apl {
+                // Already handled by is_la
+                return None;
+            }
+            if form[5] == NONE {
+                return None;
+            }
+            return form[5].into();
+        }
+        return Some(form[target as usize]);
+    } else {
+        None // Form not found
+    }
+}
+
+fn is_arabic_letter(cp: u32) -> bool {
+    cp >= AR_START && cp <= AR_END
+}
```

Comment:
It looks like you already have functionality that you could use to detect the parts of the text that uses the Arabic script. Seems like this could serve as the basis for auto-detecting parts of the text that need the specialized treatment of right-to-left change and switching of the connective forms.

Reply:
Yes, possibly for ligatures.
But specialized treatment in the case of Arabic also involves bidi text rendering, which may be better handled explicitly? I am not sure what the best choice here.

Reply:
This is a #showerthought so it might not actually be viable, but...

What you're developing here is a smart preprocessor that then allows a dumb (i.e. glyph-by-glyph) algo to render the text correctly (within the capabilities of Unifont).

For Arabic script, two things need to happen:
- reverse the order of the characters, because the script is meant to be RTL
- based on context, replace each character with the appropriate form

Reply:
Yes, so the library can expose both the string processor (smart) and the single glyph processor (dumb), correct?

Also, the introduction of `no_std` seems to break my existing implementation, and it take a while until I can get back working on this 🙏.

Reply:
No worries, I can try to take it from here!

I'll copy your branch and give you write access to it, just so that I can also work on it. :)

Reply:
Good thinking 👍 
Thanks